File: packages/go-sdk/auth_test.go
package kintsugi

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"net/http"
	"strings"
	"testing"
	"time"
)

func TestBearerAuth(t *testing.T) {
	req, _ := http.NewRequest("GET", "https://x/api/y", nil)
	BearerAuth{Token: "abc"}.Apply(req, "")
	if got := req.Header.Get("Authorization"); got != "Bearer abc" {
		t.Errorf("got %q, want %q", got, "Bearer abc")
	}
}

func TestCookieAuthNoOp(t *testing.T) {
	req, _ := http.NewRequest("GET", "https://x/api/y", nil)
	CookieAuth{}.Apply(req, "")
	if len(req.Header) != 0 {
		t.Errorf("CookieAuth should not set headers, got %v", req.Header)
	}
}

func TestAccessKeyAuthSignature(t *testing.T) {
	req, _ := http.NewRequest("POST", "https://x/api/orders/filter?ignore=me", strings.NewReader(""))
	auth := AccessKeyAuth{
		AccessKey: "ak_test",
		SecretKey: "secret",
		nowFn:     func() time.Time { return time.Unix(1700000000, 0) },
		nonceFn:   func() string { return "deadbeef12345678" },
	}
	body := `{"a":1}`
	auth.Apply(req, body)

	if got := req.Header.Get("X-Access-Key"); got != "ak_test" {
		t.Errorf("X-Access-Key: got %q want ak_test", got)
	}
	if got := req.Header.Get("X-Timestamp"); got != "1700000000" {
		t.Errorf("X-Timestamp: got %q", got)
	}
	if got := req.Header.Get("X-Nonce"); got != "deadbeef12345678" {
		t.Errorf("X-Nonce: got %q", got)
	}
	// canonical string strips query: should be /api/orders/filter (not ?ignore=me)
	canonical := "POST\n/api/orders/filter\n1700000000\ndeadbeef12345678\n" + body
	mac := hmac.New(sha256.New, []byte("secret"))
	_, _ = mac.Write([]byte(canonical))
	want := hex.EncodeToString(mac.Sum(nil))
	if got := req.Header.Get("X-Signature"); got != want {
		t.Errorf("X-Signature: got %q want %q", got, want)
	}
}

File: packages/java-sdk/src/test/java/com/kintsugi/sdk/AuthTest.java
package com.kintsugi.sdk;

import org.junit.jupiter.api.Test;

import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;
import java.net.URI;
import java.net.http.HttpRequest;
import java.security.SecureRandom;
import java.time.Clock;
import java.time.Instant;
import java.time.ZoneOffset;
import java.util.HexFormat;
import java.util.Map;

import static org.junit.jupiter.api.Assertions.*;

class AuthTest {

    @Test
    void bearerSetsAuthorization() {
        HttpRequest.Builder b = HttpRequest.newBuilder().uri(URI.create("https://x/y"));
        Auth.bearer("abc").apply(b, "GET", "/y", "");
        HttpRequest req = b.GET().build();
        assertEquals("Bearer abc", req.headers().firstValue("Authorization").orElseThrow());
    }

    @Test
    void cookieIsNoOp() {
        HttpRequest.Builder b = HttpRequest.newBuilder().uri(URI.create("https://x/y"));
        Auth.cookie().apply(b, "GET", "/y", "");
        HttpRequest req = b.GET().build();
        assertTrue(req.headers().firstValue("Authorization").isEmpty());
    }

    @Test
    void accessKeySignatureMatchesManualHmac() throws Exception {
        // pin clock + nonce so we can compute expected signature
        Clock pinned = Clock.fixed(Instant.ofEpochSecond(1700000000), ZoneOffset.UTC);
        SecureRandom seeded = new SecureRandom() {
            @Override public void nextBytes(byte[] bytes) {
                for (int i = 0; i < bytes.length; i++) bytes[i] = (byte) (0x10 + i);
            }
        };
        Auth.AccessKeyAuth auth = new Auth.AccessKeyAuth("ak_test", "secret", pinned, seeded);
        HttpRequest.Builder b = HttpRequest.newBuilder().uri(URI.create("https://x/api/y?ignore=1"));
        auth.apply(b, "POST", "/api/y?ignore=1", "{\"a\":1}");

        HttpRequest req = b.POST(HttpRequest.BodyPublishers.ofString("{\"a\":1}")).build();
        assertEquals("ak_test", req.headers().firstValue("X-Access-Key").orElseThrow());
        assertEquals("1700000000", req.headers().firstValue("X-Timestamp").orElseThrow());
        // 8-byte nonce hex = 16 chars
        String nonce = req.headers().firstValue("X-Nonce").orElseThrow();
        assertEquals(16, nonce.length());

        String canonical = "POST\n/api/y\n1700000000\n" + nonce + "\n{\"a\":1}";
        Mac mac = Mac.getInstance("HmacSHA256");
        mac.init(new SecretKeySpec("secret".getBytes(), "HmacSHA256"));
        String want = HexFormat.of().formatHex(mac.doFinal(canonical.getBytes()));
        assertEquals(want, req.headers().firstValue("X-Signature").orElseThrow());
    }
}
